### PR TITLE
ARROW-17524: [C++] Correction for fields included when reading an ORC table

### PR DESCRIPTION
**Pull request by LouisClt**
I think there is a bug in the ORC reader : when we specify the fields indexes that we want to keep, it does not work correctly. Looking at the code, it seems to be because we do "includeTypes" in lieue of "include" when setting the ORC options.
It can be problematic when we want to import an ORC table containing Union types as it will do an error at the import, even if we try not to import these specific fields.

The definitions of the corresponding ORC methods are here : 
https://github.com/apache/orc/blob/72220851cbde164a22706f8d47741fd1ad3db190/c%2B%2B/src/Options.hh#L185-L191

and 
https://github.com/apache/orc/blob/72220851cbde164a22706f8d47741fd1ad3db190/c%2B%2B/src/Options.hh#L201-L207

### CODE REVIEW

**Comment by pitrou** [comment]
@LouisClt Thanks for the report and attempt at fixing! Could you open a JIRA ticket as described above?

**Comment by LouisClt** [comment]
@pitrou Yes I can.
Here is the new Jira https://issues.apache.org/jira/browse/ARROW-17524

**Comment by LouisClt** [comment]
C Glib tests do not pass because it test the old behaviour. 
Looking a bit more in detail at the difference between "include" and "includeTypes", it appears that "includeTypes" is based on indices of the tables types. The root of the tree (meaning the whole table) is type of index 0, and when there are complex types such as structs would give other "child" types. See https://orc.apache.org/docs/types.html for more information.
"include" select the fields in the same way than the other imports : each table has a certain number of fields, (we do not take into account the "children" types), if we want to select the first field, we put "0" in the list.

I think the 2 behaviours could be understandable but the one with "include" is more coherent with the other imports. Furthermore, I do not know if there is a way in Arrow to get the list of internal ORC types, which makes selecting fields with "includeTypes" much more unreliable.

**Comment by LouisClt** [comment]
So I changed the tests to reflect the new  behaviour of selecting the fields. The job failure that remains seems to be unrelated. 
There is a decision to be made whether or not the intended behaviour of SelectFields is the previous one or this one.

**Review by pitrou — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/adapters/orc/adapter.cc [hunk 1]
```diff
@@ -411,7 +411,7 @@ class ORCFileReader::Impl {
       ARROW_RETURN_IF(*it < 0, Status::Invalid("Negative field index"));
       include_indices_list.push_back(*it);
     }
-    opts->includeTypes(include_indices_list);
+    opts->include(include_indices_list);
```
[the remark this answers is not in the document]
Comment (LouisClt) [reply cpp/src/arrow/adapters/orc/adapter.cc:414]: Indeed, it is strange that the C++ change is not tested by the C++ tests. I will try to add one test for this.

[the remark this answers is not in the document]
Comment (LouisClt) [reply cpp/src/arrow/adapters/orc/adapter.cc:414]: I added a simple test for the selection of fields in ORC. The checks that do not pass seem to be unrelated.

**Comment by pitrou** [comment]
@LouisClt Feel free to say when this is ready for another review.

**Comment by LouisClt** [comment]
Yes, this should be good now. I added a test in the ORC adapter concerning the selection of fields. Some tests did not pass, but I don't think it is related.

File: cpp/src/arrow/adapters/orc/adapter_test.cc [hunk 2]
```diff
@@ -450,6 +454,15 @@ TEST_F(TestORCWriterTrivialNoConversion, writeChunkless) {
...
 }
+TEST_F(TestORCWriterTrivialNoConversion, writeTrivialChunkAndSelectField) {
+  std::shared_ptr<Table> table = TableFromJSON(table_schema, {R"([])"});
+  std::shared_ptr<Schema> schema_selected =
+      schema({field("int8", int8()), field("int32", int32())});
+  std::shared_ptr<Table> table_selected = TableFromJSON(schema_selected, {R"([])"});
+  std::vector<int> selected_indices = {1, 3};
+  AssertTableWriteReadEqual(table, table_selected, kDefaultSmallMemStreamSize / 16,
+                            &selected_indices);
+}
```
Comment (pitrou) [inline cpp/src/arrow/adapters/orc/adapter_test.cc:465]: Thanks for the test but:
* can we also test with non-empty data?
* can we test selecting a field that's after the struct (to ensure field numbering is as expected)?

**Review by pitrou — CHANGES_REQUESTED** [review]

**Review by pitrou — APPROVED** [review]
Thanks for the update @LouisClt . This looks good to me now.

**Comment by pitrou** [comment]
CI failures are unrelated.

**Comment by LouisClt** [comment]
Good to know, thanks @pitrou

**Pull request merged**